In [2]:
# In your Jupyter Notebook

import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import plotly.express as px
from pathlib import Path
from IPython.display import display

# --- Configuration ---
# Define paths assuming this notebook is in the 'notebooks/' directory
NOTEBOOK_DIR = Path.cwd()
PROJECT_ROOT = NOTEBOOK_DIR.parent
PROCESSED_DATA_DIR = PROJECT_ROOT / "data" / "processed"

# Load the CSV file that contains the final cluster labels
CLUSTERED_FILENAME = "sale_properties_clustered.csv"
CLUSTERED_FILEPATH = PROCESSED_DATA_DIR / CLUSTERED_FILENAME

# --- Load the Data ---
try:
    df_clustered = pd.read_csv(CLUSTERED_FILEPATH)
    print("Clustered data loaded successfully.")
    print("Shape:", df_clustered.shape)
    display(df_clustered.head())
except FileNotFoundError:
    print(f"ERROR: Clustered data file not found at {CLUSTERED_FILEPATH}")

Clustered data loaded successfully.
Shape: (3002, 30)


,price_eur,barrio_encoded,adaptado_movilidad_reducida,aire_acondicionado,antiguedad,armarios_empotrados,ascensor,balcon,banos,calefaccion,...,superficie_construida,terraza,trastero,vidrios_dobles,planta_numerica,amueblado_False,amueblado_True,amueblado_nan,cluster,cluster_label
0,9500000.0,2.711931e+06,0.0,1.0,0.214609,1.0,1.0,1.0,5.0,1.0,...,499.0,0.0,0.0,0.0,1.000000,0.0,1.0,0.0,4,Premium Reformado
1,840000.0,7.521907e+05,0.0,0.0,0.000000,0.0,0.0,1.0,2.0,1.0,...,109.0,0.0,0.0,0.0,2.000000,0.0,1.0,0.0,4,Premium Reformado
2,3600000.0,1.985104e+06,0.0,1.0,0.000000,0.0,1.0,1.0,5.0,1.0,...,379.0,1.0,0.0,1.0,3.000000,0.0,0.0,1.0,0,Piso Señorial Clásico
3,285000.0,2.620667e+05,0.0,1.0,2.000000,0.0,0.0,0.0,1.0,1.0,...,60.0,0.0,0.0,0.0,2.284866,0.0,0.0,1.0,1,Apartamento Estándar
4,539000.0,4.553499e+05,1.0,1.0,2.000000,1.0,1.0,0.0,2.0,1.0,...,127.0,0.0,1.0,1.0,2.000000,0.0,0.0,1.0,3,Lujo Moderno (Full Equip)


In [3]:
print("--- Cluster Size Distribution ---")
cluster_counts = df_clustered['cluster'].value_counts().sort_index()
print(cluster_counts)

# Visualize the distribution
fig = px.bar(
    cluster_counts, 
    x=cluster_counts.index, 
    y=cluster_counts.values,
    title="Number of Properties per Cluster",
    labels={'x': 'Cluster ID', 'y': 'Number of Properties'}
)
fig.show()

--- Cluster Size Distribution ---
cluster
0     660
1    1365
2       2
3     388
4     587
Name: count, dtype: int64


In [4]:
print("--- Analysis of Numerical Features by Cluster ---")

# List of key numerical features to analyze
numerical_features = [
    'price_eur',
    'superficie_construida',
    'habitaciones',
    'banos',
    'planta_numerica',
    'antiguedad', # This is now numerical (ordinal)
    'conservacion', # This is now numerical (ordinal)
    'barrio_encoded', # The target-encoded value
]

# Group by cluster and calculate the mean and median for each feature
# Using .agg() allows us to calculate multiple statistics at once
cluster_summary_numerical = df_clustered.groupby('cluster')[numerical_features].agg(['mean', 'median', 'count'])

# Display the summary table
# We'll round the values for easier reading
display(cluster_summary_numerical.round(2))

--- Analysis of Numerical Features by Cluster ---


price_eur                  superficie_construida               \
               mean     median count                  mean median count   
cluster                                                                   
0        1596959.58  1247000.0   660                166.58  137.0   660   
1         626263.69   335000.0  1365                 96.15   77.0  1365   
2        1797500.00  1797500.0     2                441.00  441.0     2   
3        1908280.41  1280000.0   388                256.81  200.0   388   
4        1508718.31  1199000.0   587                132.99  115.0   587   

        habitaciones              banos  ... planta_numerica antiguedad  \
                mean median count  mean  ...           count       mean   
cluster                                  ...                              
0               3.11    3.0   660  2.58  ...             660       0.43   
1               2.57    3.0  1365  1.57  ...            1365       0.41   
2               5.50    5.5     2  5.00  ...               2       1.00   
3               4.25    4.0   388  3.74  ...             388       1.98   
4               2.70    3.0   587  2.47  ...             587       0.35   

                     conservacion              barrio_encoded              \
        median count         mean median count           mean      median   
cluster                                                                     
0         0.05   660         1.22   1.00   660     1543640.62  1650219.56   
1         0.21  1365         1.19   1.00  1365      747432.94   415265.71   
2         1.00     2         1.00   1.00     2     3038955.36  3038955.36   
3         2.00   388         1.46   1.32   388     1375750.03  1174431.82   
4         0.00   587         1.96   2.00   587     1634669.52  1715335.71   

               
        count  
cluster        
0         660  
1        1365  
2           2  
3         388  
4         587  

[5 rows x 24 columns]

In [5]:
print(cluster_summary_numerical)

            price_eur                  superficie_construida               \
                 mean     median count                  mean median count   
cluster                                                                     
0        1.596960e+06  1247000.0   660            166.579302  137.0   660   
1        6.262637e+05   335000.0  1365             96.149099   77.0  1365   
2        1.797500e+06  1797500.0     2            441.000000  441.0     2   
3        1.908280e+06  1280000.0   388            256.806671  200.0   388   
4        1.508718e+06  1199000.0   587            132.985893  115.0   587   

        habitaciones                  banos  ... planta_numerica antiguedad  \
                mean median count      mean  ...           count       mean   
cluster                                      ...                              
0           3.110724    3.0   660  2.576369  ...             660   0.429760   
1           2.573740    3.0  1365  1.571289  ...            1365   

In [6]:
print("\n--- Analysis of Boolean (Amenity) Features by Cluster ---")

# List of key boolean features
boolean_features = [
    'ascensor',
    'piscina',
    'garaje',
    'terraza',
    'trastero',
    'exterior',
    # Include the one-hot encoded columns for 'amueblado'
    'amueblado_True',
    'amueblado_False',
]

# Group by cluster and calculate the mean (percentage) for each boolean feature
cluster_summary_boolean = df_clustered.groupby('cluster')[boolean_features].mean()

# Display the summary table (values will be from 0.0 to 1.0)
display((cluster_summary_boolean * 100).round(2).style.background_gradient(cmap='viridis').format("{:.2f}%"))


--- Analysis of Boolean (Amenity) Features by Cluster ---


,ascensor,piscina,garaje,terraza,trastero,exterior,amueblado_True,amueblado_False
cluster,,,,,,,,
0,93.18%,2.73%,22.58%,41.82%,31.21%,74.85%,0.00%,0.00%
1,33.26%,1.61%,11.21%,27.62%,11.06%,29.01%,0.07%,0.00%
2,50.00%,50.00%,50.00%,50.00%,50.00%,50.00%,0.00%,100.00%
3,42.27%,80.67%,84.02%,53.61%,57.47%,47.16%,5.15%,0.00%
4,78.36%,1.70%,11.24%,28.28%,16.01%,52.30%,100.00%,0.00%


In [7]:
print("\n--- Visualizing Price Distribution by Cluster ---")

fig = px.box(
    df_clustered,
    x='cluster',
    y='price_eur',
    color='cluster',
    title='Distribution of Property Prices by Cluster',
    labels={'cluster': 'Cluster ID', 'price_eur': 'Price (€)'},
    points="all" # Show all underlying data points
)
fig.update_layout(xaxis_type='category') # Treat cluster IDs as categories
print("\n--- Visualizing Mean Feature Values by Cluster ---")fig.show()

SyntaxError: invalid syntax (1955113913.py, line 13)

In [ ]:
print("\n--- Visualizing Mean Feature Values by Cluster ---")

# Displaying the data in a formatted table for clarity before plotting.
print("\n--- Summary Table: Mean Feature Values by Cluster ---")
# Using the Pandas Styler for a more professional-looking table output
styled_means = cluster_means.style.format({
    'superficie_construida': '{:,.2f}', # Format with comma for thousands and 2 decimal places
    'banos': '{:.2f}',                  # Format to 2 decimal places
    'habitaciones': '{:.2f}'            # Format to 2 decimal places
}).set_caption("Average Feature Values per Cluster").hide(axis='index')

# The 'display()' function provides a richer rendering in Jupyter Notebooks
display(styled_means)

# Compare average surface area and number of bathrooms
features_to_plot = ['superficie_construida', 'banos', 'habitaciones']
cluster_means = df_clustered.groupby('cluster')[features_to_plot].mean().reset_index()

# We need to "melt" the DataFrame to make it suitable for a grouped bar chart
df_melted = cluster_means.melt(id_vars='cluster', value_vars=features_to_plot, var_name='Feature', value_name='Average Value')

fig = px.bar(
    df_melted,
    x='cluster',
    y='Average Value',
    color='Feature',
    barmode='group', # Group bars for the same cluster side-by-side
    title='Comparison of Average Feature Values Across Clusters',
    labels={'cluster': 'Cluster ID'}
)
fig.update_layout(xaxis_type='category')
fig.show()


--- Visualizing Mean Feature Values by Cluster ---

--- Summary Table: Mean Feature Values by Cluster ---


cluster,superficie_construida,banos,habitaciones
0,166.58,2.58,3.11
1,96.15,1.57,2.57
2,441.00,5.00,5.50
3,256.81,3.74,4.25
4,132.99,2.47,2.70


## Key recommendation:
Cluster 2, with only two properties, should be treated as an outlier or anomaly cluster. While its characteristics are interesting (ultra-large, ultra-expensive), it's too small to be a representative group. For the next step (training a classifier), we'll exclude these two properties from the training set, as a model can't learn meaningful patterns from just two examples and might even be negatively affected by them.

## Proposed Labels and Profiles for Each Cluster


### Cluster 1: Standard & Affordable Apartments
This is the largest group (1365 properties), representing the most common type of property on the market.

#### Defining Characteristics:

*   **Price & Value**: Has the lowest average price (~€626k) and is located in neighborhoods with the lowest average property value (barrio_encoded).
*   **Size**: Has the smallest average size (~96 m²) and fewest rooms/bathrooms.
*   **Amenities**: Lacks most luxury amenities. It has the lowest percentage of elevators (33%), pools (1.6%), and garages (11%).
*   **Condition**: Generally older properties in "good condition" but typically not recently renovated.
*   **Persona**: This cluster represents the core of the Madrid housing market: standard, smaller apartments in less central or less expensive neighborhoods. They are functional homes but generally lack modern luxuries like pools, garages, or guaranteed elevators, which is typical for a large portion of the city's building stock.


### Cluster 0: Classic, Large City Center Apartments
This is a significant group of 660 properties that represents a very classic Madrid property type.

#### Defining Characteristics:

*   **Price & Value**: High price (~€1.6M) and located in high-value neighborhoods.
*   **Size**: Quite large (~167 m²) with 3+ rooms and 2.5+ bathrooms on average.
*   **Amenities**: Very high chance of having an elevator (93%) and being an exterior-facing unit (75%). However, it has a very low percentage of pools and garages.
*   **Condition**: Tends to be in older buildings (antiguedad mean is very low).
*   **Persona**: This cluster perfectly describes the classic, spacious apartments found in the established, central districts of Madrid (often called a "piso señorial"). These are desirable properties in older, well-maintained buildings that have elevators but typically lack the space for amenities like private pools or large garages, which are rare in historic city centers.


### Cluster 4: Premium Renovated & Furnished Apartments
This group of 587 properties represents a high-end, "turn-key" segment of the market.

#### Defining Characteristics:

*   **Price & Value**: High price (~€1.5M), located in the most expensive neighborhoods (highest barrio_encoded value of the main groups).
*   **Condition**: These are, on average, the oldest properties but have the best conservation status, strongly indicating they have been recently renovated.
*   **Amenities**: High chance of an elevator (78%). A defining feature is that 100% of these properties are furnished (amueblado_True).
*   **Size**: Medium-sized (~133 m²), smaller than the "classic" apartments in Cluster 0.
*   **Persona**: This cluster represents apartments in prime, historic neighborhoods that have been fully renovated to a high standard and are sold ready to be lived in. They are perfect for buyers looking for a premium, hassle-free property in a top location, or for investors seeking a high-yield rental property.

### Cluster 3: Modern Luxury Homes & Penthouses
This group of 388 properties clearly represents the modern, high-end luxury market.

#### Defining Characteristics:

*   **Price & Value**: The highest average price of all major clusters (~€1.9M).
*   **Size**: Very large (~257 m²) with the most rooms and bathrooms (~4.25 and ~3.7 respectively).
*   **Amenities**: This cluster is defined by its luxury amenities. It has by far the highest percentage of pools (81%) and garages (84%), along with a high likelihood of having a terrace and storage room (trastero).
*   **Condition**: These are the newest properties on average (antiguedad score is highest).
*   **Persona**: This cluster represents modern luxury. These are likely penthouses, duplexes, or houses/chalets in newer, exclusive residential complexes (urbanizaciones) where amenities like pools, private garages, and large terraces are standard features.

### Summary of Property Clusters (Sales Data)

| ID Cluster | Etiqueta | Características Clave                                                                        |
| :--------- | :-------------------------- | :------------------------------------------------------------------------------------------- |
| **0** | Piso Señorial Clásico       | Grande, precio elevado, finca antigua, con ascensor, pero sin piscina/garaje.                |
| **1** | Apartamento Estándar        | El más económico, tamaño reducido y menos extras. El apartamento "típico" de Madrid.         |
| **2** | **Propiedad Singular (Outlier)** | Anomalía. Extremadamente grande y caro. (Se recomienda excluir del entrenamiento del clasificador). |
| **3** | Lujo Moderno (Full Equip)   | El más caro, grande, obra nueva o reciente, y equipado con piscina, garaje y terraza.       |
| **4** | Premium Reformado           | Precio elevado, ubicación prime, finca antigua pero en excelente estado (reformado), y amueblado. |